README.md
# barcode-reader

## Barcode reader

Barcode-reader is a library for reading a variety of barcode types in images.

[![Lint](https://github.com/antoinehumbert/barcode-reader/actions/workflows/lint.yml/badge.svg)](https://github.com/antoinehumbert/barcode-reader/actions/workflows/lint.yml)
[![Test](https://github.com/antoinehumbert/barcode-reader/actions/workflows/test.yml/badge.svg)](https://github.com/antoinehumbert/barcode-reader/actions/workflows/test.yml)
[![codecov](https://codecov.io/gh/antoinehumbert/barcode-reader/branch/develop/graph/badge.svg?token=90FVQSXYS9&flag=GHA_Ubuntu)](https://codecov.io/gh/antoinehumbert/barcode-reader)
[![Build + Deploy](https://github.com/antoinehumbert/barcode-reader/actions/workflows/build.yml/badge.svg)](https://github.com/antoinehumbert/barcode-reader/actions/workflows/build.yml)
[![PyPI](https://img.shields.io/pypi/v/barcode-reader.svg)](https://pypi.org/project/barcode-reader/)
[![Documentation Status](https://readthedocs.org/projects/barcode-reader/badge/?version=latest)](https://barcode-reader.readthedocs.io/en/latest/?badge=latest)

## Overview

The barcode-reader library provides functionalities for reading a variety of barcode types from images and return useful
information such as the barcode data but also its coordinates.

## More information

- [Documentation](https://barcode-reader.readthedocs.io/)
- Contribute
  - [Issues](https://github.com/antoinehumbert/barcode-reader/issues)
  - [Pull requests](https://github.com/antoinehumbert/barcode-reader/pulls)
- [Changelog](https://github.com/antoinehumbert/barcode-reader/blob/master/CHANGES.rst)


CHANGES.rst
Changelog (bercode-reader)
==========================

0.1.0 (unreleased)
------------------

Initial version


tests/conftest.py
from pathlib import Path

import pkg_resources
import pytest
import atexit
# noinspection PyProtectedMember
from barcode_reader._zxing import _executors


@pytest.fixture
def resources():
    yield Path(pkg_resources.resource_filename("tests", "resources"))


@pytest.fixture(scope="session", autouse=True)
def _handle_executors():
    """
    This fixture ensure that Zxing executors are shutdown at the end of tests execution which is necessary for
    pytest-cov to get results from the _zxing module parts executed in subprocesses
    """
    yield
    _executors.shutdown()


tests/barcode_reader/test_utils.py
import sys

import cv2
import numpy as np
from PIL import Image

# noinspection PyProtectedMember
from barcode_reader.utils import detect_barcodes


class TestReader:

    DEBUG = False
    # Some results depend on the running platform. May some some rounding difference !
    WIN = sys.platform == "win32"
    DARWIN = sys.platform == "darwin"

    def test_detect_qrcode(self, resources):
        quiet_dist = 10
        min_area = 12000
        multi_dm_image = resources / "qrcode.png"
        with Image.open(multi_dm_image).convert("RGB") as image:
            rectangles = [
                np.int0(cv2.boxPoints(((center_x, center_y), (width + 2 * quiet_dist, height + 2 * quiet_dist), angle)))
                for (center_x, center_y), (width, height), angle in detect_barcodes(image, quiet_dist, min_area)
            ]
            assert [[tuple(p) for p in rect] for rect in rectangles] == [
                [(24, 274), (24, 23), (274, 23), (274, 274)],
            ]
            if self.DEBUG:
                cv_im = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
                cv2.drawContours(cv_im, rectangles, -1, (0, 255, 0), 3)
                image.frombytes(cv2.cvtColor(cv_im, cv2.COLOR_BGR2RGB))
                image.show()

    def test_detect_qrcode_multi(self, resources):
        quiet_dist = 10
        min_area = 12000
        multi_dm_image = resources / "qrcode_multi.png"
        with Image.open(multi_dm_image).convert("RGB") as image:
            rectangles = [
                np.int0(cv2.boxPoints(((center_x, center_y), (width + 2 * quiet_dist, height + 2 * quiet_dist), angle)))
                for (center_x, center_y), (width, height), angle in detect_barcodes(image, quiet_dist, min_area)
            ]
            assert [[tuple(p) for p in rect] for rect in rectangles] == [
                [(87, 1615), (87, 1364), (338, 1364), (338, 1615)],
                [(2187, 1578), (2253, 1335), (2496, 1400), (2431, 1644)],
                [(484, 1400), (727, 1335), (793, 1578), (549, 1644)],
                [(1744, 1535), (1870, 1317), (2088, 1443), (1962, 1661)],
                [(892, 1443), (1110, 1318), (1236, 1535), (1018, 1661)],
                [(1313, 1489), (1490, 1312), (1667, 1489), (1490, 1666)],
                [(87, 1191), (87, 939), (337, 939), (337, 1191)],
                [(484, 975), (728, 910), (793, 1153), (550, 1219)],
                [(2188, 1153), (2254, 909), (2497, 975), (2432, 1218)],
                [(1745, 1110), (1871, 892), (2089, 1018), (1963, 1236)],
                [(893, 1018), (1111, 892), (1237, 1110), (1019, 1236)],
                [(1314, 1064), (1491, 887), (1668, 1064), (1491, 1241)],
                [(85, 763), (85, 512), (337, 512), (337, 763)],
                [(2188, 728), (2253, 484), (2497, 550), (2431, 793)],
                [(483, 550), (727, 484), (792, 728), (549, 793)],
                [(1744, 685), (1870, 467), (2088, 593), (1962, 811)],
                [(892, 593), (1110, 467), (1236, 685), (1018, 810)],
                [(1313, 639), (1490, 462), (1667, 639), (1490, 816)],
                [(87, 337), (87, 85), (337, 86), (337, 337)],
                [(2187, 301), (2252, 58), (2496, 123), (2430, 367)],
                [(483, 123), (726, 57), (792, 301), (548, 366)],
                [(1743, 258), (1869, 40), (2087, 166), (1961, 384)],
                [(892, 166), (1109, 40), (1235, 258), (1017, 384)],
                [(1312, 212), (1489, 35), (1666, 212), (1489, 389)],
            ]
            if self.DEBUG:
                cv_im = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
                cv2.drawContours(cv_im, rectangles, -1, (0, 255, 0), 3)
                image.frombytes(cv2.cvtColor(cv_im, cv2.COLOR_BGR2RGB))
                image.show()

    def test_detect_datamatrix(self, resources):
        quiet_dist = 10
        min_area = 12000
        multi_dm_image = resources / "datamatrix.png"
        with Image.open(multi_dm_image).convert("RGB") as image:
            rectangles = [
                np.int0(cv2.boxPoints(((center_x, center_y), (width + 2 * quiet_dist, height + 2 * quiet_dist), angle)))
                for (center_x, center_y), (width, height), angle in detect_barcodes(image, quiet_dist, min_area)
            ]
            assert [[tuple(p) for p in rect] for rect in rectangles] == [
                [(-4, 282), (-4, -2), (280, -2), (280, 282)],
            ]
            if self.DEBUG:
                cv_im = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
                cv2.drawContours(cv_im, rectangles, -1, (0, 255, 0), 3)
                image.frombytes(cv2.cvtColor(cv_im, cv2.COLOR_BGR2RGB))
                image.show()

    def test_detect_datamatrix_multi(self, resources):
        quiet_dist = 10
        min_area = 12000
        multi_dm_image = resources / "datamatrix_multi.png"
        with Image.open(multi_dm_image).convert("RGB") as image:
            rectangles = [
                np.int0(cv2.boxPoints(((center_x, center_y), (width + 2 * quiet_dist, height + 2 * quiet_dist), angle)))
                for (center_x, center_y), (width, height), angle in detect_barcodes(image, quiet_dist, min_area)
            ]
            assert [[tuple(p) for p in rect] for rect in rectangles] == [
                [(59, 1561), (59, 1276), (343, 1276), (343, 1561)],
                [(2056, 1521), (2130, 1246), (2405, 1320), (2332, 1595)],
                [(432, 1319), (707, 1245), (781, 1520), (506, 1594)],
                [(1630, 1472), (1773, 1226), (2019, 1368), (1877, 1615)],
                [(818, 1368), (1065, 1225), (1207, 1472), (961, 1614)],
                [(1218, 1420), (1419, 1219), (1619, 1420), (1419, 1621)],
                [(61, 1156), (61, 871), (345, 871), (345, 1156)],
                [(2057, 1114), (2131, 838), (2406, 912), (2332, 1187)],
                [(434, 912), (709, 838), (783, 1114), (508, 1187)],
                [(1631, 1065), (1774, 818), (2020, 961), (1878, 1207)],
                [(820, 961), (1066, 818), (1209, 1065), (962, 1207)],
                [(1219, 1013), (1420, 812), (1621, 1013), (1420, 1213)],
                [(61, 752), (61, 467), (345, 467), (345, 752)],
                [(435, 508), (710, 434), (784, 709), (508, 783)],
                [(2059, 708), (2132, 433), (2408, 507), (2334, 782)],
                [(821, 556), (1067, 414), (1210, 660), (963, 803)],
                [(1633, 660), (1775, 413), (2022, 556), (1879, 802)],
                [(1221, 608), (1422, 407), (1622, 608), (1422, 809)],
                [(59, 345), (59, 61), (343, 61), (343, 345)],
                [(2058, 304), (2132, 29), (2407, 102), (2333, 378)],
                [(433, 102), (708, 29), (782, 304), (507, 378)],
                [(1632, 255), (1774, 9), (2021, 151), (1878, 398)],
                [(819, 151), (1066, 9), (1208, 255), (962, 398)],
                [(1219, 204), (1420, 3), (1621, 204), (1420, 404)],
            ]
            if self.DEBUG:
                cv_im = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
                cv2.drawContours(cv_im, rectangles, -1, (0, 255, 0), 3)
                image.frombytes(cv2.cvtColor(cv_im, cv2.COLOR_BGR2RGB))
                image.show()


src/barcode_reader/_dmtx.py
import ctypes
from collections import namedtuple
from ctypes import cast, string_at
from functools import partial

# noinspection PyProtectedMember
from pylibdmtx.pylibdmtx import (
    ENCODING_SIZE_PREFIX, ENCODING_SIZE_NAMES, ENCODING_SCHEME_PREFIX, ENCODING_SCHEME_NAMES, _encoder, Encoded,
    _decoded_matrix_region, Rect, _pixel_data, _decoder, _image, _PACK_ORDER, _region
)
from pylibdmtx.pylibdmtx_error import PyLibDMTXError
# noinspection PyProtectedMember
from pylibdmtx.wrapper import (
    c_ubyte_p, dmtxTimeAdd, dmtxTimeNow, dmtxMatrix3VMultiplyBy, dmtxDecodeSetProp, DmtxProperty, DmtxUndefined,
    DmtxVector2, DmtxSymbolSize, DmtxScheme, dmtxEncodeSetProp, dmtxEncodeDataMatrix, dmtxImageGetProp
)

Point = namedtuple('Point', 'x y')
Decoded = namedtuple('Decoded', 'data rect polygon')


def _decode_region(decoder, region, corrections, shrink, height):
    """Decodes and returns the value in a region.

    Args:
        region (DmtxRegion):

    Yields:
        Decoded or None: The decoded value.
    """
    with _decoded_matrix_region(decoder, region, corrections) as msg:
        if msg:
            # Coordinates
            p00 = DmtxVector2()
            p10 = DmtxVector2(1.0, 0)
            p11 = DmtxVector2(1.0, 1.0)
            p01 = DmtxVector2(0.0, 1.0)
            dmtxMatrix3VMultiplyBy(p00, region.contents.fit2raw)
            dmtxMatrix3VMultiplyBy(p10, region.contents.fit2raw)
            dmtxMatrix3VMultiplyBy(p11, region.contents.fit2raw)
            dmtxMatrix3VMultiplyBy(p01, region.contents.fit2raw)
            top_left = Point(int((shrink * p01.X) + 0.5), int(height - (shrink * p01.Y) - 0.5))
            top_right = Point(int((shrink * p11.X) + 0.5), int(height - (shrink * p11.Y) - 0.5))
            bottom_right = Point(int((shrink * p10.X) + 0.5), int(height - (shrink * p10.Y) - 0.5))
            bottom_left = Point(int((shrink * p00.X) + 0.5), int(height - (shrink * p00.Y) - 0.5))
            polygon = [top_left, top_right, bottom_right, bottom_left]
            x_min = min(p.x for p in polygon)
            x_max = max(p.x for p in polygon)
            y_min = min(p.y for p in polygon)
            y_max = max(p.y for p in polygon)
            return Decoded(
                string_at(msg.contents.output),
                Rect(x_min, y_min, x_max - x_min, y_max - y_min),
                polygon
            )
        else:
            return None


def decode(image, timeout=None, gap_size=None, shrink=1, shape=None,
           deviation=None, threshold=None, min_edge=None, max_edge=None,
           corrections=None, max_count=None):
    """Decodes datamatrix barcodes in `image`.

    Args:
        image: `numpy.ndarray`, `PIL.Image` or tuple (pixels, width, height)
        timeout (int): milliseconds
        gap_size (int):
        shrink (int):
        shape (int):
        deviation (int):
        threshold (int):
        min_edge (int):
        max_edge (int):
        corrections (int):
        max_count (int): stop after reading this many barcodes. `None` to read
            as many as possible.

    Returns:
        :obj:`list` of :obj:`Decoded`: The values decoded from barcodes.
    """
    dmtx_timeout = None
    if timeout:
        now = dmtxTimeNow()
        dmtx_timeout = dmtxTimeAdd(now, timeout)

    if max_count is not None and max_count < 1:
        raise ValueError('Invalid max_count [{0}]'.format(max_count))

    pixels, width, height, bpp = _pixel_data(image)

    results = []
    with _image(
        cast(pixels, c_ubyte_p), width, height, _PACK_ORDER[bpp]
    ) as img:
        with _decoder(img, shrink) as decoder:
            properties = [
                (DmtxProperty.DmtxPropScanGap, gap_size),
                (DmtxProperty.DmtxPropSymbolSize, shape),
                (DmtxProperty.DmtxPropSquareDevn, deviation),
                (DmtxProperty.DmtxPropEdgeThresh, threshold),
                (DmtxProperty.DmtxPropEdgeMin, min_edge),
                (DmtxProperty.DmtxPropEdgeMax, max_edge)
            ]

            # Set only those properties with a non-None value
            for prop, value in ((p, v) for p, v in properties if v is not None):
                dmtxDecodeSetProp(decoder, prop, value)

            if not corrections:
                corrections = DmtxUndefined

            while True:
                with _region(decoder, dmtx_timeout) as region:
                    # Finished file or ran out of time before finding another
                    # region
                    if not region:
                        break
                    else:
                        # Decoded
                        res = _decode_region(
                            decoder, region, corrections, shrink, height
                        )
                        if res:
                            results.append(res)

                            # Stop if we've reached maximum count
                            if max_count and len(results) == max_count:
                                break

    return results


def encode(data, scheme=None, size=None):
    """
    Encodes `data` in a DataMatrix image.

    For now bpp is the libdmtx default which is 24

    Args:
        data: bytes instance
        scheme: encoding scheme - one of `ENCODING_SCHEME_NAMES`, or `None`.
            If `None`, defaults to 'Ascii'.
        size: image dimensions - one of `ENCODING_SIZE_NAMES`, or `None`.
            If `None`, defaults to 'ShapeAuto'.

    Returns:
        Encoded: with properties `(width, height, bpp, pixels)`.
        You can use that result to build a PIL image:

            Image.frombytes('RGB', (width, height), pixels)

    """

    size = size if size else 'ShapeAuto'
    size_name = '{0}{1}'.format(ENCODING_SIZE_PREFIX, size)
    if not hasattr(DmtxSymbolSize, size_name):
        raise PyLibDMTXError(
            'Invalid size [{0}]: should be one of {1}'.format(
                size, ENCODING_SIZE_NAMES
            )
        )
    size = getattr(DmtxSymbolSize, size_name)

    scheme = scheme if scheme else 'Ascii'
    scheme_name = '{0}{1}'.format(
        ENCODING_SCHEME_PREFIX, scheme
    )
    if not hasattr(DmtxScheme, scheme_name):
        scheme_name = '{0}{1}'.format(
            ENCODING_SCHEME_PREFIX, scheme.capitalize()
        )
    if not hasattr(DmtxScheme, scheme_name):
        raise PyLibDMTXError(
            'Invalid scheme [{0}]: should be one of {1}'.format(
                scheme, ENCODING_SCHEME_NAMES
            )
        )
    scheme = getattr(DmtxScheme, scheme_name)

    with _encoder() as encoder:
        dmtxEncodeSetProp(encoder, DmtxProperty.DmtxPropScheme, scheme)
        dmtxEncodeSetProp(encoder, DmtxProperty.DmtxPropSizeRequest, size)

        if dmtxEncodeDataMatrix(encoder, len(data), cast(data, c_ubyte_p)) == 0:
            raise PyLibDMTXError(
                'Could not encode data, possibly because the image is not '
                'large enough to contain the data'
            )

        w, h, bpp = map(
            partial(dmtxImageGetProp, encoder[0].image),
            (
                DmtxProperty.DmtxPropWidth, DmtxProperty.DmtxPropHeight,
                DmtxProperty.DmtxPropBitsPerPixel
            )
        )
        size = w * h * bpp // 8
        pixels = cast(
            encoder[0].image[0].pxl, ctypes.POINTER(ctypes.c_ubyte * size)
        )
        return Encoded(
            width=w, height=h, bpp=bpp, pixels=ctypes.string_at(pixels, size)
        )


src/barcode_reader/_zbar.py
"""
This module is a fork of pyzbar.pyzbar with a modification in _decode_symbols function so that the returned polygon for
a barcode always starts with the point being the top-left corner of the barcode and coordinates are in clockwise order.

It also adds an additional functions for correct decoding of barcode data.
"""
import operator
from collections import deque
from ctypes import cast, c_void_p, string_at, c_uint, c_char_p

# noinspection PyPackageRequirements
from pyzbar.locations import bounding_box, convex_hull
# noinspection PyProtectedMember,PyPackageRequirements
from pyzbar.pyzbar import _image_scanner, _pixel_data, _FOURCC, _symbols_for_image, _RANGEFN, _image, Decoded
# noinspection PyPackageRequirements
from pyzbar.pyzbar_error import PyZbarError
# noinspection PyProtectedMember,PyPackageRequirements
from pyzbar.wrapper import (
    zbar_image_scanner_set_config, zbar_symbol_get_quality, zbar_image_set_format, zbar_image_set_size,
    zbar_image_set_data, zbar_scan_image, zbar_symbol_get_data, zbar_symbol_get_orientation, zbar_symbol_get_loc_size,
    zbar_symbol_get_loc_x, zbar_symbol_get_loc_y, zbar_symbol_xml, ZBarConfig, ZBarSymbol, ZBarOrientation
)


def _decode_zbar_data(data):
    """
    Zbar returns data in utf-8 encoding. But, when reading the barcode, it decoded it with the sjis encoding whereas
    the default should be iso-8859-1 for barcode producers that do not use utf-8 encoding.

    For this reason, we try to decode the zbar returned data by replacing sjis encoding by iso-8859-1. But if this
    does not work, we assume that original encoding was utf-8 and do not re-encode/decode zbar data

    :param bytes data: the zbar read data (utf-8 encoded)
    :return: the decoded data
    :rtype: str
    """
    data_str = data.decode("utf-8")
    # For unknown reason, when a QR-Code is encoded in iso-8859-1 with 2 and only 2 non-ascii characters, zbar decode it
    # using sjis or big5 encoding instead of iso-8859-1. So we try here to correct this.
    try:
        try:
            data = data_str.encode("sjis")
        except UnicodeEncodeError:
            data = data_str.encode("big5")
    except UnicodeEncodeError:
        pass  # It was not decoded using sjis nor big5 encoding, so this is ok
    else:
        iso_8859_1 = data.decode("iso-8859-1")
        # If string contains 2 and only 2 non-ascii characters, it was probably iso-8859-1 encoded and not sjis
        if len(iso_8859_1.encode("ascii", errors="ignore")) == len(data) - 2:
            data_str = iso_8859_1
    return data_str


def _decode_symbols(symbols, xml=False):
    """Generator of decoded symbol information.

    Args:
        symbols: iterable of instances of `POINTER(zbar_symbol)`

    Yields:
        Decoded: decoded symbol
    """
    for symbol in symbols:
        data = _decode_zbar_data(string_at(zbar_symbol_get_data(symbol)))
        # The 'type' int in a value in the ZBarSymbol enumeration
        symbol_type = ZBarSymbol(symbol.contents.type).name
        polygon = [
            (zbar_symbol_get_loc_x(symbol, index), zbar_symbol_get_loc_y(symbol, index))
            for index in _RANGEFN(zbar_symbol_get_loc_size(symbol))
        ]
        x, y = polygon[0]
        polygon = deque(convex_hull(polygon))
        # change polygon points from counterclockwise order to clockwise order
        polygon.reverse()
        # Reorder convex polygon points so that first point is the closest to first point of original polygon (i.e. the
        # top left point of the barcode)
        x_y_dists = [(idx, (px - x) ** 2 + (py - y) ** 2) for idx, (px, py) in enumerate(polygon)]
        min_dist_idx = min(x_y_dists, key=operator.itemgetter(1))[0]
        polygon.rotate(-min_dist_idx)
        polygon = list(polygon)
        orientation = ZBarOrientation(zbar_symbol_get_orientation(symbol))
        quality = zbar_symbol_get_quality(symbol)

        if xml:
            xml_len_p = c_uint()
            xml_buf = c_char_p()
            item = zbar_symbol_xml(symbol, xml_buf, xml_len_p)
        else:
            item = Decoded(
                data=data,
                type=symbol_type,
                rect=bounding_box(polygon),
                polygon=polygon,
                orientation=orientation,
                quality=quality,
            )
        yield item


# noinspection PyIncorrectDocstring
def decode(image, symbols=None, xml=False):
    """Decodes datamatrix barcodes in `image`.

    Args:
        image: `numpy.ndarray`, `PIL.Image` or tuple (pixels, width, height)
        symbols: iter(ZBarSymbol) the symbol types to decode; if `None`, uses
            `zbar`'s default behaviour, which is to decode all symbol types.

    Returns:
        :obj:`list` of :obj:`Decoded`: The values decoded from barcodes.
    """
    pixels, width, height = _pixel_data(image)

    results = []
    with _image_scanner() as scanner:
        if symbols:
            # Disable all but the symbols of interest
            disable = set(ZBarSymbol).difference(symbols)
            for symbol in disable:
                zbar_image_scanner_set_config(
                    scanner, symbol, ZBarConfig.CFG_ENABLE, 0
                )
            # I think it likely that zbar will detect all symbol types by
            # default, in which case enabling the types of interest is
            # redundant but it seems sensible to be over-cautious and enable
            # them.
            for symbol in symbols:
                zbar_image_scanner_set_config(
                    scanner, symbol, ZBarConfig.CFG_ENABLE, 1
                )
        with _image() as img:
            zbar_image_set_format(img, _FOURCC['L800'])
            zbar_image_set_size(img, width, height)
            zbar_image_set_data(img, cast(pixels, c_void_p), len(pixels), None)
            decoded = zbar_scan_image(scanner, img)
            if decoded < 0:
                raise PyZbarError('Unsupported image format')
            else:
                results.extend(_decode_symbols(_symbols_for_image(img), xml))

    return results


src/barcode_reader/_zxing/__init__.py
import operator
from collections import namedtuple, deque
from concurrent.futures import ProcessPoolExecutor

# noinspection PyPackageRequirements
import cv2
# noinspection PyPackageRequirements
import jpype
import numpy as np
from imforge.crop import crop
from pkg_resources import resource_filename

from barcode_reader.utils import detect_barcodes

Point = namedtuple("Point", ["x", "y"])
Rect = namedtuple('Rect', ['left', 'top', 'width', 'height'])
Decoded = namedtuple('Decoded', ["data", "type", "rect", "polygon"])
_ZxingResult = namedtuple("ZxingResult", ["data", "format", "result_points"])

# zxing functions requiring the JVM are launch in other processes to avoid interaction with possible other libraries
# using jpype and needing a JVM with different configuration
_CLASS_PATHS = (
    resource_filename(__name__, "core-3.4.1.jar"),
    resource_filename(__name__, "javase-3.4.1.jar"),
)


def _start_zxing_jvm():
    # noinspection PyUnresolvedReferences
    jpype.startJVM(classpath=_CLASS_PATHS)


class ZxingDecoder:

    def __init__(self):
        # Binary image generation
        self.DataBufferByte = jpype.JClass("java.awt.image.DataBufferByte")
        self.Raster = jpype.JClass("java.awt.image.Raster")
        self.BufferedImage = jpype.JClass("java.awt.image.BufferedImage")
        self.BinaryBitmap = jpype.JClass("com.google.zxing.BinaryBitmap")
        self.HybridBinarizer = jpype.JClass("com.google.zxing.common.HybridBinarizer")
        self.BufferedImageLuminanceSource = jpype.JClass("com.google.zxing.client.j2se.BufferedImageLuminanceSource")
        # exceptions
        self.NotFoundException = jpype.JClass("com.google.zxing.NotFoundException")
        # readers
        self.multi_reader = jpype.JClass("com.google.zxing.MultiFormatReader")()
        self.qr_reader = jpype.JClass("com.google.zxing.multi.qrcode.QRCodeMultiReader")()
        self.pdf417_reader = jpype.JClass("com.google.zxing.pdf417.PDF417Reader")()
        # hints
        # noinspection PyPep8Naming
        ArrayList = jpype.JClass("java.util.ArrayList")
        # noinspection PyPep8Naming
        BarcodeFormat = jpype.JClass("com.google.zxing.BarcodeFormat")
        # noinspection PyPep8Naming
        DecodeHintType = jpype.JClass("com.google.zxing.DecodeHintType")
        self.qr_hints = {DecodeHintType.TRY_HARDER: True}
        self.pdf417_hints = {DecodeHintType.TRY_HARDER: True}
        self.other_hints = {
            DecodeHintType.POSSIBLE_FORMATS: ArrayList([
                BarcodeFormat.AZTEC, BarcodeFormat.CODABAR, BarcodeFormat.CODE_39, BarcodeFormat.CODE_93,
                BarcodeFormat.CODE_128, BarcodeFormat.DATA_MATRIX, BarcodeFormat.EAN_8, BarcodeFormat.EAN_13,
                BarcodeFormat.ITF, BarcodeFormat.MAXICODE, BarcodeFormat.RSS_14, BarcodeFormat.RSS_EXPANDED,
                BarcodeFormat.UPC_E,
                # BarcodeFormat.PDF_417, BarcodeFormat.QR_CODE,  # use specific readers for QR_CODE and PDF_417
            ])
        }

    def decode(self, image, quiet_dist=10, min_area=12000):
        """
        Decode barcodes of given image (must be in L mode). This function is intended for execution in child process.
        The Zxing JVM must have been started with the _start_zxing_jvm function before calling this function.

        :param PIL.Image.Image image: the image (in L mode) to decode
        :param int quiet_dist: the quiet distance in pixels (i.e. blank zone around barcodes)
        :param int min_area: the min area in pixels x pixels a barcode must have
        :return: the list of decoding results
        :rtype: list[barcode_reader._zxing.Decoded]
        """
        image = image.convert("L")
        im = np.array(image)
        bitmap = self.get_binary_bitmap(im)
        results = self.decode_qr(bitmap)
        results.extend(self.decode_pdf417(bitmap))
        for result in results:
            cv2.fillConvexPoly(im, np.int0(result.result_points), color=255)
        image.frombytes(im)
        all_points = [pt for result in results for pt in result.result_points]
        for center, (width, height), angle in detect_barcodes(image, quiet_dist, min_area):
            rectangle = deque(
                (round(p[0]), round(p[1]))
                for p in cv2.boxPoints((center, (width + 2 * quiet_dist, height + 2 * quiet_dist), angle))
            )
            # Check if rectangle contains an already identified barcode. If so, we skip it
            if any(cv2.pointPolygonTest(np.array(rectangle), pt, False) >= 0 for pt in all_points):
                continue
            # change rectangle and angle to have the top-left corner first
            top_left_idx = min(
                ((idx, x * x + y * y) for idx, (x, y) in enumerate(rectangle)), key=operator.itemgetter(1)
            )[0]
            rectangle.rotate(-top_left_idx)
            rectangle = list(rectangle)
            if top_left_idx == 0:
                angle -= 90
            # elif min_dist_idx == 1:
            #    the rectangle is correctly oriented. No change
            elif top_left_idx == 2:
                angle += 90
            elif top_left_idx == 3:
                angle += 180
            mat = cv2.getRotationMatrix2D((width / 2 + quiet_dist, height / 2 + quiet_dist), -angle, 1)
            top_left = rectangle[0]
            mat[0][2] = top_left[0]
            mat[1][2] = top_left[1]
            im = np.array(crop(image, rectangle, fillcolor=255))
            for zxing_result in self.decode_other(self.get_binary_bitmap(im)):
                results.append(
                    _ZxingResult(
                        data=zxing_result.data,
                        format=zxing_result.format,
                        result_points=cv2.transform(
                            np.array([zxing_result.result_points], dtype=np.float32), mat
                        ).tolist()[0]
                    )
                )
        decoded_results = []
        for result in results:
            polygon = deque(Point(x=round(x), y=round(y)) for x, y in result.result_points)
            polygon.rotate(-1)
            polygon = list(polygon)
            x_min = min(p.x for p in polygon)
            x_max = max(p.x for p in polygon)
            y_min = min(p.y for p in polygon)
            y_max = max(p.y for p in polygon)
            decoded_results.append(
                Decoded(
                    data=result.data,
                    type=result.format,
                    rect=Rect(left=x_min, top=y_min, width=x_max - x_min, height=y_max - y_min),
                    polygon=polygon,
                )
            )
        return decoded_results

    def get_binary_bitmap(self, image):
        """
        Return binary bitmap for given grayscale (numpy array) image

        :param numpy.array image: the grayscale image to convert
        :return: the binary bitmap (a ``com.google.zxing.BinaryBitmap`` instance)
        """
        height, width = image.shape
        buf_image = self.BufferedImage(width, height, self.BufferedImage.TYPE_BYTE_GRAY)
        buf_image.setData(
            self.Raster.createBandedRaster(
                self.DataBufferByte(jpype.JArray.of(image.flatten(), dtype=jpype.JByte), width * height),
                width, height, width, [0], [0], None
            )
        )
        return self.BinaryBitmap(self.HybridBinarizer(self.BufferedImageLuminanceSource(buf_image)))

    def decode_qr(self, bitmap):
        """
        Decode QR-Codes in given binary bitmap

        :param bitmap: the bitmap to decode (a ``com.google.zxing.BinaryBitmap`` instance)
        :return: the list of read QR-Codes
        :rtype: list[barcode_reader._zxing._ZxingResult]
        """
        results = []
        # Detect multiple QR-Codes
        reader = self.qr_reader
        reader.reset()
        try:
            qr_results = reader.decodeMultiple(bitmap, self.qr_hints)
        except self.NotFoundException:
            pass
        else:
            for result in qr_results:
                result_points = [(float(p.getX()), float(p.getY())) for p in result.getResultPoints()]
                data = str(result.getText())
                barcode_format = str(result.getBarcodeFormat())
                results.append(_ZxingResult(data=data, format=barcode_format, result_points=result_points))
        return results

    def decode_pdf417(self, bitmap):
        """
        Decode PDF417 Codes in given binary bitmap

        :param bitmap: the bitmap to decode (a ``com.google.zxing.BinaryBitmap`` instance)
        :return: the list of read PDF417 Codes
        :rtype: list[barcode_reader._zxing._ZxingResult]
        """
        results = []
        # Detect multiple PDF-417
        reader = self.pdf417_reader
        reader.reset()
        try:
            pdf417_results = reader.decodeMultiple(bitmap, self.pdf417_hints)
        except self.NotFoundException:
            pass
        else:
            for result in pdf417_results:
                result_points = [(float(p.getX()), float(p.getY())) for p in result.getResultPoints()]
                data = str(result.getText())
                barcode_format = str(result.getBarcodeFormat())
                results.append(_ZxingResult(data=data, format=barcode_format, result_points=result_points))
        return results

    def decode_other(self, bitmap):
        """
        Decode **one** barcode (not QR-Code or PDF417) in given binary bitmap

        :param bitmap: the bitmap to decode (a ``com.google.zxing.BinaryBitmap`` instance)
        :return: the list of read PDF417 Codes
        :rtype: list[barcode_reader._zxing._ZxingResult]
        """
        results = []
        # Finally, search for one code with the MultiFormatReader
        reader = self.multi_reader
        reader.reset()
        try:
            any_result = reader.decode(bitmap, self.other_hints)
        except self.NotFoundException:
            pass
        else:
            result_points = [(float(p.getX()), float(p.getY())) for p in any_result.getResultPoints()]
            data = str(any_result.getText())
            barcode_format = str(any_result.getBarcodeFormat())
            if barcode_format == "DATA_MATRIX":
                # Zxing usually decode datamatrix using iso-8859-1 charset. However, it may be encoded in utf-8.
                # So, we try to re-decode data using utf-8 encoding. If it works, we consider this was the correct
                # encoding. If not, we keep the originally decoded data from Zxing
                try:
                    data = data.encode("iso-8859-1").decode("utf-8")
                except (UnicodeEncodeError, UnicodeDecodeError):
                    pass
            results.append(_ZxingResult(data=data, format=barcode_format, result_points=result_points))
        return results


def _init_process():
    global _zxing_decoder
    _start_zxing_jvm()
    _zxing_decoder = ZxingDecoder()


_zxing_decoder = None
_executors = ProcessPoolExecutor(initializer=_init_process)


def _decode(image):
    # noinspection PyUnresolvedReferences
    return _zxing_decoder.decode(image)


def decode(image):
    """
    Decode barcodes of given image

    :param PIL.Image.Image image: the image to decode
    :return: list of found barcodes
    :rtype: list[barcode_reader._zxing.Decoded]
    """
    return _executors.submit(_decode, image).result()
